core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java

```diff
@@ -611,7 +613,17 @@ private void applyInitializingShard(final ClusterState state, IndexService index
         }
 
         indexShard.startRecovery(nodes.getLocalNode(), sourceNode, recoveryTargetService,
-            new RecoveryListener(shardRouting, indexService), repositoriesService);
+            new RecoveryListener(shardRouting, indexService), repositoriesService, (type, mapping) -> {
+                try {
+                    nodeServicesProvider.getClient().admin().indices().preparePutMapping()
```

Comment:
this is tricky - how do we secure this for example? . I wonder if we should go through a simpler route where the master copies the mapping from the source index to the target index  when resolving INDEX_SHRINK_SOURCE_NAME . The IndexShard can just verify that the source and target mapping are identical, and if not fail the allocation?